runtime/neumann/src/fee_handler.rs

```diff
@@ -0,0 +1,171 @@
+where
+	T: pallet_transaction_payment::Config,
+	C: Currency<<T as frame_system::Config>::AccountId>,
+	C::Balance: From<MC::Balance>,
+	C::PositiveImbalance: Imbalance<
+		<C as Currency<<T as frame_system::Config>::AccountId>>::Balance,
+		Opposite = C::NegativeImbalance,
+	>,
+	C::NegativeImbalance: Imbalance<
+		<C as Currency<<T as frame_system::Config>::AccountId>>::Balance,
+		Opposite = C::PositiveImbalance,
+	>,
+	MC::CurrencyId: From<TokenId>,
+	MC::Balance: From<C::Balance>,
+	MC: MultiCurrency<<T as frame_system::Config>::AccountId>,
+	OU: OnUnbalanced<NegativeImbalanceOf<C, T>>,
+	TA: Get<T::AccountId>,
+	FCP: CallParser<CallOf<T>>,
+{
+	type LiquidityInfo = Option<NegativeImbalanceOf<C, T>>;
+	type Balance = <C as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	/// Withdraw the prxedicted fee from the transaction origin.
+	///
+	/// Note: The `fee` already includes the `tip`.
+	fn withdraw_fee(
+		who: &T::AccountId,
+		call: &T::Call,
+		_info: &DispatchInfoOf<T::Call>,
+		fee: Self::Balance,
+		tip: Self::Balance,
+	) -> Result<Self::LiquidityInfo, TransactionValidityError> {
+		if fee.is_zero() {
+			return Ok(None)
+		}
+
+		let call_information = FCP::fee_information(call.clone());
+
+		if call_information.token_id != NATIVE_TOKEN_ID && call_information.xcm_data == None ||
+			call_information.asset_metadata == None
+		{
+			return Err(TransactionValidityError::Invalid(InvalidTransaction::Payment))
+		}
+
+		if call_information.token_id == NATIVE_TOKEN_ID {
+			let withdraw_reason = if tip.is_zero() {
+				WithdrawReasons::TRANSACTION_PAYMENT
+			} else {
+				WithdrawReasons::TRANSACTION_PAYMENT | WithdrawReasons::TIP
+			};
+
+			match C::withdraw(who, fee, withdraw_reason, ExistenceRequirement::KeepAlive) {
+				Ok(imbalance) => Ok(Some(imbalance)),
+				Err(_) => Err(InvalidTransaction::Payment.into()),
+			}
+		} else {
+			use sp_runtime::SaturatedConversion;
+			let currency_id = call_information.token_id.into();
+			let foreign_fee = call_information
+				.asset_metadata
+				.ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?
+				.additional
+				.convert_fee_into_foreign(fee.saturated_into())
+				.ok_or(TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			// orml_tokens doesn't provide withdraw that allows setting existence so prevent
+			// withdraw if they don't have enough to avoid reaping
+			MC::ensure_can_withdraw(
+				currency_id,
+				who,
+				foreign_fee
+					.saturating_add(MC::minimum_balance(currency_id).saturated_into())
+					.saturated_into(),
+			)
+			.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			MC::withdraw(currency_id, who, foreign_fee.saturated_into())
+				.map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+
+			MC::deposit(currency_id, &TA::get(), foreign_fee.saturated_into()) // treasury account
```

Comment:
@irsal Do we want foreign funds to go to a different treasury accont.

Reply:
what was the existing one? or are we talking about the actual treasury?